packages/api/src/car.js

```diff
@@ -303,48 +300,24 @@ async function carStat (carBytes) {
   if (!rawRootBlock) {
     throw new Error('missing root block')
   }
-  if (blocks === 1) {
-    const decoder = decoders.find(d => d.code === rootCid.code)
-    // if we can't decode, we can't check this...
-    if (decoder) {
-      const rootBlock = new Block({ cid: rootCid, bytes: rawRootBlock.bytes, value: decoder.decode(rawRootBlock.bytes) })
-      const numLinks = Array.from(rootBlock.links()).length
-      // if the root block has links, then we should have at least 2 blocks in the CAR
-      if (numLinks > 0) {
-        throw new Error('CAR must contain at least one non-root block')
-      }
+  const decoder = decoders.find(d => d.code === rootCid.code)
+  if (decoder) {
+    const rootBlock = new Block({ cid: rootCid, bytes: rawRootBlock.bytes, value: decoder.decode(rawRootBlock.bytes) })
+    const links = Array.from(rootBlock.links())
+    // if the root block has links, then we should have at least 2 blocks in the CAR
+    if (blocks === 1 && links.length > 0) {
+      throw new Error('CAR must contain at least one non-root block')
+    }
+    // get the size of the full dag for this root, even if we only have a partial CAR.
+    if (rootCid.code === pb.code) {
+      size = cumulativeSize(rootBlock, links)
     }
   }
   return { size, blocks }
 }
 
-/**
- * Returns the DAG size of the CAR but only if the graph is complete.
- * @param {Uint8Array} carBytes
- */
-async function getDagSize (carBytes) {
-  const reader = await CarReader.fromBytes(carBytes)
-  const [rootCid] = await reader.getRoots()
-
-  const getBlock = async cid => {
-    const rawBlock = await reader.get(cid)
-    if (!rawBlock) throw new Error(`missing block for ${cid}`)
-    const { bytes } = rawBlock
-    const decoder = decoders.find(d => d.code === cid.code)
-    if (!decoder) throw new Error(`missing decoder for ${cid.code}`)
-    return new Block({ cid, bytes, value: decoder.decode(bytes) })
-  }
-
-  const getSize = async cid => {
-    const block = await getBlock(cid)
-    let size = block.bytes.byteLength
-    for (const [, cid] of block.links()) {
-      size += await getSize(cid)
-    }
-    return size
-  }
-
-  return getSize(rootCid)
+function cumulativeSize (block, links) {
```

Comment:
Can we add jsdoc here? Can you point me where Tsize is? I could not find it in multiformats

Reply:
I pulled this from `ipfs-core` here: https://github.com/ipfs/js-ipfs/blob/b5488cb3ef5ab7ea24b8b2275ffa8cfa4b7b7ed1/packages/ipfs-core/src/components/files/stat.js#L121

Reply:
Ok, I could not find where this type is documented but I found that it is added in unixfs https://github.com/ipfs/js-ipfs-unixfs/search?q=Tsize

Reply:
this is the most canonical thing I could find https://ipld.io/specs/codecs/dag-pb/spec/#serial-format

Reply:
thanks, I would suggest we add a comment with that link to be more clear

Reply:
also https://github.com/ipld/js-dag-pb/blob/998e0b2a501de6e844d1b0e9a86034b474427080/src/interface.ts#L8-L12

Reply:
yeah, this isn't doing what i think it's doing! Thanks for the nudge.